### ISSUE TITLE
transform_prox_AoAs_to_blade_AoAs  Problem when Using ToAs Extracted from Ansys for Constant Speed Cases

### ISSUE THREAD

**Issue by JustinSmith66** [Bladesight/Bladesight#62]
### Summary

- When processing Times of Arrival (ToAs) extracted from Ansys for constant-speed cases, the function `transform_prox_AoAs_to_blade_AoAs` often produces NaN or inf values in blades_AoAs.
- For my code, this appears to be triggered by slightly imprecise probe spacing (e.g., not exactly 9.666... degrees, and was instead a rounded variant, for the BTT test-bench case) and seems related to numerical issues in bin assignment.

### Observed Symptoms

- **`blades_AoAs` contains many `NaN` values**:
    - The first two probes may produce valid results.
    - From probe 2 onwards, values become `NaN` if spacing is not exactly `9.666...` recurring.
- Printing `blade_median_AoAs` inside `transform_prox_AoAs_to_blade_AoAs` often shows **`inf`** in the second index.
- This happens when **no bin is assigned** to certain blades, resulting in empty dataframes (`df_bin`) and failed median calculations.

### Suspected Cause

- The `WHERE` clause in the SQL that creates `df_bin` for the **last blade** sometimes returns no rows.
- Reasons this may occur:
    1. **Poor bin separation**:
        - `calculate_Q` and the `optimal_d_theta` optimisation loop may produce `optimal_bin_edges` where one bin is empty.
        - Example: `(optimal_bin_edges[1], optimal_bin_edges[2])` falls in a gap in the AoA data.
    2. **Skewed AoA distribution**:
        - All AoAs fall into a single blade’s bin after ToA → AoA conversion, leaving others empty.
- In these cases, the `duckdb` median query runs on an empty dataframe, returning `NULL`, which the code converts to `np.inf`.

### Why This Feels like a Numerical Issue

- Even with evenly distributed data from constant-speed cases, small deviations in **probe spacing** appear to prevent proper bin assignment.
- This suggests the bin edge calculations are numerically fragile.

### Proposed Fix

- Replace the bin-based approach with **direct clustering** (e.g., K-Means) on AoA values.
- This avoids the problem of empty bins by letting the clustering algorithm determine group boundaries.
- I know that it is undesirable to have additional dependencies for `bladesight`. However, I could not think of an alternative.

```
// ...existing code...
import pandas as pd
import polars as pl

from bladesight.btt.aoa import calculate_Q
from bladesight.utils import _get_dataframe_library_preference

def transform_prox_AoAs_to_blade_AoAs(
    df_prox: Union[pd.DataFrame, pl.DataFrame],
    B: int,
    d_theta_increments: int = 200, # This parameter is no longer used but kept for API consistency
    spurious_pulse_policy : Literal['drop', 'keep'] = 'drop'
) -> List[Union[pd.DataFrame, pl.DataFrame]]:
    """This function takes a DataFrame containing the AoA values of a proximity probe,
    and returns a list of DataFrame, each containing the AoA values of a single blade.

    Args:
        df_prox (Union[pd.DataFrame, pl.DataFrame]): The dataframe containing the AoA values
            of the proximity probe.
        B (int): The number of blades.
        d_theta_increments (int, optional): This parameter is no longer used.
        spurious_pulse_policy (Literal['drop', 'keep'], optional): The policy to handle
            spurious pulses. Defaults to 'drop'.

    Returns:
        List[Union[pd.DataFrame, pl.DataFrame]]: A list of dataframes, each containing the AoA
        values of a single blade.
    """
    return_type = _get_dataframe_library_preference()
    
    # Use K-Means for robust clustering of AoA data
    try:
        from sklearn.cluster import KMeans
    except ImportError:
        raise ImportError("scikit-learn is required for this function. Please install it with 'pip install scikit-learn'")

    arr_aoas = duckdb.execute("SELECT AoA FROM df_prox").fetchnumpy()["AoA"]
    
    # K-Means works on Cartesian coordinates, so convert angles
    X = np.vstack([np.cos(arr_aoas), np.sin(arr_aoas)]).T
    
    kmeans = KMeans(n_clusters=B, random_state=0, n_init='auto').fit(X)
    labels = kmeans.labels_
    
    # Add the cluster labels to the dataframe
    if return_type == 'pd':
        df_prox_labeled = duckdb.execute("SELECT * FROM df_prox").df()
        df_prox_labeled['blade_label'] = labels
    else:
        df_prox_labeled = duckdb.execute("SELECT * FROM df_prox").pl()
        df_prox_labeled = df_prox_labeled.with_columns(pl.Series("blade_label", labels))

    blade_dfs = []
    blade_median_AoAs = []
    
    # Group by the new labels
    for i in range(B):
        if return_type == 'pd':
            df_bin = df_prox_labeled[df_prox_labeled['blade_label'] == i]
        else:
            df_bin = df_prox_labeled.filter(pl.col('blade_label') == i)
        
        blade_dfs.append(df_bin)
        
        # Calculate median from the populated bin
        med = duckdb.execute("SELECT MEDIAN(AoA) FROM df_bin").fetchone()[0]
        blade_median_AoAs.append(med if med is not None else np.nan)

    blade_median_AoAs = np.array(blade_median_AoAs, dtype=float)
    
    # Handle cases where a cluster might be empty, though unlikely with K-Means
    if np.isnan(blade_median_AoAs).any():
        print("Warning: K-Means resulted in an empty cluster.")
        # Replace nan with inf to maintain sorting behavior
        blade_median_AoAs = np.nan_to_num(blade_median_AoAs, nan=np.inf)

    blade_order = np.argsort(blade_median_AoAs)
    blade_dfs = [blade_dfs[i] for i in blade_order]
    
    if spurious_pulse_policy == 'keep':
        return blade_dfs
    
    # Drop revolutions that has more than one occurrence of a blade
    blade_dfs_to_return = []
    query = """
        SELECT * EXCLUDE(blade_label) FROM df
        QUALIFY row_number() OVER (PARTITION BY n ORDER BY ToA) = 1
        ORDER BY n, ToA
    """
    for _, df in enumerate(blade_dfs):
        if df.shape[0] == 0:
            # Append empty dataframe if a cluster was empty
            blade_dfs_to_return.append(df.drop(columns=['blade_label'], errors='ignore'))
            continue

        if return_type == 'pd':
            blade_dfs_to_return.append(duckdb.execute(query).df())
        else:
            blade_dfs_to_return.append(duckdb.execute(query).pl())
            
    return blade_dfs_to_return

def get_rotor_blade_AoAs(
// ...existing code...
```

**Comment by JustinSmith66** [comment]
# Comment when this issue receives attention:

I noticed that my proposed method gives slightly less accurate AoAs than the old one. The AoAs appear to become less and less accurate as the signal continues. I can not confirm the severity, but I want to investigate this issue further. 

I suspect this is because the proposed fix using K-Means correctly solves the problem of empty bins and crashes. Still, it produces less accurate Angle of Arrivals (AoAs) because it fundamentally changes the logic from a physics-informed model to a purely statistical one.

The original binning method, despite its fragility, is more accurate because it correctly assumes the blades are equally spaced angularly. The K-Means approach makes no such assumption and simply groups data points by proximity.

# Recommendation: A Hybrid Approach
The issue isn't the binning logic itself, but its fragility. Instead of replacing it entirely, you could make it more robust.

A powerful solution would be to use K-Means to find an excellent starting guess for the original algorithm.

Run K-Means First: Perform the K-Means clustering as you proposed. This gives you a robust, albeit potentially inaccurate, grouping.

Calculate Cluster Centers: For each of the B clusters found by K-Means, calculate the median angle. This will give you B median angles, e.g., [theta_1, theta_2, ...., B]

Find the Optimal Offset: These median angles give a near-perfect estimate of the rotor's angular position. You can use them to calculate a much better initial d_theta to feed into the original binning algorithm, virtually guaranteeing that it finds the correct alignment without failing.

Run the Binning Algorithm: With this high-quality starting point, the original, physically-constrained binning algorithm can perform the final, accurate assignment of AoAs to each blade.

This hybrid approach combines the robustness of K-Means with the physical accuracy of the original binning model, giving you the best of both worlds.

## A Possible Solution
- The only problem I am having is finding super accurate AoAs. I am proposing the below as a starting point, but I think further improvements can be made to make the AoAs much more accurate and have the BTT match simulation data perfectly.
- All the previous edits are pertaining to the below code. I am testing this with my own data and as I notice an issue, and fix it, I post the latest code here.

```
import warnings
def transform_prox_AoAs_to_blade_AoAs(
    df_prox: Union[pd.DataFrame, pl.DataFrame],
    B: int,
    kmeans_n_init: int = 50,
    kmeans_max_iter: int = 5000,
    kmeans_tol: float = 1e-4,
    d_theta_increments: int = 20000,
    spurious_pulse_policy: Literal['drop', 'keep'] = 'drop',
    min_cluster_fraction: float = 0.02,
    debug: bool = False
) -> List[Union[pd.DataFrame, pl.DataFrame]]:
    """
    Split proximity probe AoAs into per-blade AoAs using a hybrid strategy.

    Flow:
        0. Basic validation & NaN filtering.
        1. Attempt K-Means clustering on the unit circle representation of angles.
            - Reject result if: any cluster too small, or angular span collapsed.
        2. If K-Means invalid or fails, fall back to brute force sweep over d_theta.
        3. Build angular bins (accounting for wrap at 2π) and extract rows per bin.
        4. Sort resulting blade bins by median AoA (stable ordering).
        5. Optionally de-duplicate multiple hits per revolution (keep earliest).

    Warnings emitted when:
        - K-Means fails or is rejected (will use brute force).
        - Brute force path is taken.
        - Any output bin is empty (can cause downstream NaNs after pivot).

    Parameters
    ----------
    df_prox : DataFrame (pandas or polars)
        Must contain at least columns: AoA, ToA, n (revolution index).
    B : int
        Number of blades.
    kmeans_n_init, kmeans_max_iter, kmeans_tol : K-Means tuning.
    d_theta_increments : int
        Resolution of brute force sweep range for offset if fallback used.
    spurious_pulse_policy : 'drop' | 'keep'
        If 'drop', keep only first hit per revolution per blade.
    min_cluster_fraction : float
        Minimum fraction of total samples allowed per cluster.
    debug : bool
        Verbose console output.

    Returns
    -------
    list of DataFrames
        One per blade, possibly empty if data sparse.
    """
    return_type = _get_dataframe_library_preference()

    # ---- 0. Guard: empty dataset -------------------------------------------------
    if df_prox.shape[0] == 0:
        if debug:
            print("[transform] Empty input DataFrame.")
        if return_type == 'pd':
            return [pd.DataFrame(columns=df_prox.columns)] * B
        else:
            return [pl.DataFrame(schema=df_prox.schema)] * B

    # Register in DuckDB once (re-register later after potential mutation)
    duckdb.register("df_prox", df_prox)

    # Pull AoAs & ToAs (and clean)
    arr = duckdb.execute("SELECT AoA, ToA FROM df_prox").fetchnumpy()
    aoa_raw = arr["AoA"]
    toa_raw = arr["ToA"] if "ToA" in arr else None

    # Mask invalids (NaNs / infs)
    mask = np.isfinite(aoa_raw)
    if toa_raw is not None:
        mask &= np.isfinite(toa_raw)

    aoa = aoa_raw[mask]
    toa = toa_raw[mask] if toa_raw is not None else None

    if aoa.size < B:
        warnings.warn(
            f"[transform] Not enough valid AoAs ({aoa.size}) for {B} blades. Returning empty bins.",
            RuntimeWarning
        )
        if debug:
            print(f"[transform] Not enough valid AoAs ({aoa.size}) for {B} blades.")
        if return_type == 'pd':
            return [pd.DataFrame(columns=df_prox.columns)] * B
        else:
            return [pl.DataFrame(schema=df_prox.schema)] * B

    # Optional: sort by time (stabilizes brute force and bin boundary consistency)
    if toa is not None:
        order = np.argsort(toa)
        aoa = aoa[order]
        toa = toa[order]

    # Quick variance check (extremely low spread can doom clustering)
    if np.std(aoa) < 1e-6:
        warnings.warn(
            "[transform] Very low AoA variance detected; clustering may be unreliable.",
            RuntimeWarning
        )

    # ---- 1. Attempt K-Means clustering -------------------------------------------
    optimal_d_theta = None  # Will be set by K-Means or brute force

    def circular_mean(angles: np.ndarray) -> float:
        """Mean of angles on unit circle (robust to wrap)."""
        if len(angles) == 0:
            return 0.0
        return np.arctan2(np.sum(np.sin(angles)), np.sum(np.cos(angles)))

    def brute_force_offset(a: np.ndarray) -> float:
        """
        Exhaustive grid search over a plausible offset window.
        Returns offset minimizing Q (dispersion metric from calculate_Q).
        """
        d_thetas = np.linspace(-0.5 * np.pi / B, 1.5 * np.pi / B, d_theta_increments)
        best_Q = np.inf
        best_dt = 0.0
        for dth in d_thetas:
            Q, _ = calculate_Q(a, dth, B)
            if Q < best_Q:
                best_Q = Q
                best_dt = dth
        return best_dt

    tried_kmeans = False
    if aoa.size >= B:
        try:
            from sklearn.cluster import KMeans
            tried_kmeans = True
            # Represent each angle on the unit circle
            X = np.vstack([np.cos(aoa), np.sin(aoa)]).T
            kmeans = KMeans(
                n_clusters=B,
                random_state=0,
                n_init=kmeans_n_init,
                max_iter=kmeans_max_iter,
                tol=kmeans_tol,
            ).fit(X)
            labels = kmeans.labels_

            # --- Validate populations
            counts = np.array([np.sum(labels == i) for i in range(B)])
            min_allowed = max(1, int(min_cluster_fraction * aoa.size))
            if debug:
                print(f"[transform] KMeans cluster counts={counts}, min_allowed={min_allowed}")
            if (counts < min_allowed).any():
                raise ValueError("Cluster population below minimum threshold")

            # --- Means per cluster (angular)
            cluster_means = np.array([circular_mean(aoa[labels == i]) for i in range(B)])
            cluster_means = np.sort(cluster_means)

            # --- Angular span check to detect collapse
            span = np.mod(cluster_means[-1] - cluster_means[0], 2 * np.pi)
            min_expected_span = 0.6 * (2 * np.pi * (B - 1) / B)
            if debug:
                print(f"[transform] Angular span={span:.4f}, required>={min_expected_span:.4f}")
            if span < min_expected_span:
                raise ValueError("Insufficient angular spread (clusters collapsed)")

            # Map means to ideal evenly spaced positions to get global offset
            ideal = np.arange(B) * (2 * np.pi / B)
            offsets = cluster_means - ideal
            optimal_d_theta = circular_mean(offsets)

            if debug:
                print(f"[transform] KMeans accepted. offset={optimal_d_theta:.6f} rad")

        except Exception as e:
            # Emit a warning (only if K-Means attempted)
            warnings.warn(
                f"[transform] K-Means rejected ({type(e).__name__}: {e}). Falling back to brute force.",
                RuntimeWarning
            )
            optimal_d_theta = None

    # ---- 2. Brute force fallback -------------------------------------------------
    if optimal_d_theta is None:
        optimal_d_theta = brute_force_offset(aoa)
        warnings.warn(
            f"[transform] Using brute force offset search. d_theta={optimal_d_theta:.6f} rad",
            RuntimeWarning
        )
        if debug:
            print(f"[transform] Brute force chosen. d_theta={optimal_d_theta:.6f} rad (tried_kmeans={tried_kmeans})")

    # ---- 3. Build bin edges ------------------------------------------------------
    # Normalize base offset into [0, 2π)
    base = np.mod(optimal_d_theta, 2 * np.pi)
    # Create B contiguous edges without wrapping the upper bound (preserves ordering)
    edges = base + np.arange(B + 1) * (2 * np.pi / B)
    if debug:
        print(f"[transform] Bin edges: {edges}")

    # Re-register original DataFrame in case external code mutated it
    duckdb.register("df_prox", df_prox)

    # ---- 4. Assign rows to bins (SQL with wrap) ----------------------------------
    blade_dfs: List[Union[pd.DataFrame, pl.DataFrame]] = []
    medians = []

    for i in range(B):
        lower = edges[i]
        upper = edges[i + 1]

        query = f"""
            WITH norm AS (
                SELECT
                    *,
                    CASE
                        WHEN AoA < {base} THEN AoA + 2*PI()
                        ELSE AoA
                    END AS AoA_norm
                FROM df_prox
            )
            SELECT * EXCLUDE(AoA_norm)
            FROM norm
            WHERE AoA_norm >= {lower} AND AoA_norm < {upper}
            ORDER BY ToA
        """

        if return_type == 'pd':
            df_bin = duckdb.execute(query).df()
        else:
            df_bin = duckdb.execute(query).pl()

        duckdb.register("df_bin", df_bin)
        med = duckdb.execute("SELECT MEDIAN(AoA) FROM df_bin").fetchone()[0]
        medians.append(med if med is not None else np.inf)
        blade_dfs.append(df_bin)

    # ---- 5. Order blades by median AoA for consistency ---------------------------
    medians = np.array(medians, dtype=float)
    order = np.argsort(medians)
    blade_dfs = [blade_dfs[j] for j in order]

    if debug:
        print(f"[transform] Median AoAs (sorted order): {medians[order]}")

    if spurious_pulse_policy == 'keep':
        # Return raw per-bin DataFrames (possibly with duplicate revolutions)
        return blade_dfs

    # ---- 6. De-duplicate: keep earliest hit per revolution per blade -------------
    cleaned: List[Union[pd.DataFrame, pl.DataFrame]] = []
    qualify_query = """
        SELECT * FROM df_current
        QUALIFY row_number() OVER (PARTITION BY n ORDER BY ToA) = 1
        ORDER BY n, ToA
    """
    for df_part in blade_dfs:
        if df_part is None or (hasattr(df_part, "shape") and df_part.shape[0] == 0):
            cleaned.append(df_part)
            continue
        duckdb.register("df_current", df_part)
        if return_type == 'pd':
            cleaned.append(duckdb.execute(qualify_query).df())
        else:
            cleaned.append(duckdb.execute(qualify_query).pl())

    # ---- 7. Post-check for empty bins --------------------------------------------
    empty_bins = [i for i, d in enumerate(cleaned) if (hasattr(d, "shape") and d.shape[0] == 0)]
    if empty_bins:
        warnings.warn(
            f"[transform] Empty blade bins at indices {empty_bins}. Downstream pivot may yield NaNs.",
            RuntimeWarning
        )
        if debug:
            print(f"[transform] Empty bins detail indices={empty_bins}")

    return cleaned


```
